plugins/providers/gcs/provider.go

```diff
@@ -0,0 +1,253 @@
+	}
+
+	client, err := p.getGCSClient(creds)
+	if err != nil {
+		return fmt.Errorf("error in getting new client: %w", err)
+	}
+
+	identity := fmt.Sprintf("%s:%s", a.AccountType, a.AccountID) // identity is AccountType : AccountID, eg: "serviceAccount:test@email.com"
+	if a.Resource.Type == ResourceTypeBucket {
+		b := new(Bucket)
+		if err := b.fromDomain(a.Resource); err != nil {
+			return fmt.Errorf("from Domain func error: %w", err)
+		}
+		for _, p := range permissions {
+			role := iam.RoleName(string(p))
+			if err := client.GrantBucketAccess(context.TODO(), *b, identity, role); err != nil {
+				if errors.Is(err, ErrPermissionAlreadyExists) {
+					return nil
+				}
+				return err
+			}
+		}
+		return nil
+	}
+	return ErrInvalidResourceType
+}
+
+func (p *Provider) RevokeAccess(pc *domain.ProviderConfig, a *domain.Appeal) error {
+	if err := validateProviderConfigAndAppealParams(pc, a); err != nil {
+		return fmt.Errorf("invalid provider/appeal config: %w", err)
+	}
+
+	permissions, err := getPermissions(pc.Resources, a)
+	if err != nil {
+		return fmt.Errorf("error in getting permissions: %w", err)
+	}
+
+	var creds Credentials
+	if err := mapstructure.Decode(pc.Credentials, &creds); err != nil {
+		return fmt.Errorf("error in decoding credentials%w", err)
+	}
+
+	if err := creds.Decrypt(p.crypto); err != nil {
+		return fmt.Errorf("error in decrypting credentials%w", err)
+	}
+
+	client, err := p.getGCSClient(creds)
+	if err != nil {
+		return fmt.Errorf("error in getting new client: %w", err)
+	}
+
+	user := a.AccountID
+	userType := a.AccountType
+	identity := fmt.Sprintf("%s:%s", userType, user)
+	ctx := context.TODO()
+	if a.Resource.Type == ResourceTypeBucket {
+		b := new(Bucket)
+		if err := b.fromDomain(a.Resource); err != nil {
+			return fmt.Errorf("from Domain func error: %w", err)
+		}
+		for _, p := range permissions {
+			var role iam.RoleName = iam.RoleName(string(p))
+			if err := client.RevokeBucketAccess(ctx, *b, identity, role); err != nil {
+				if errors.Is(err, ErrPermissionAlreadyExists) {
+					return nil
+				}
+				return err
+			}
+		}
+		return nil
+	}
+	return ErrInvalidResourceType
+}
+
+func (p *Provider) GetRoles(pc *domain.ProviderConfig, resourceType string) ([]*domain.Role, error) {
+	return provider.GetRoles(pc, resourceType)
+}
+
+func (p *Provider) GetAccountTypes() []string {
+	return []string{AccountTypeUser, AccountTypeServiceAccount, AccountTypeGroup, AccountTypeDomain}
+}
+
+func validateProviderConfigAndAppealParams(pc *domain.ProviderConfig, a *domain.Appeal) error {
+	if pc == nil {
+		return ErrNilProviderConfig
+	}
+	if a == nil {
+		return ErrNilAppeal
+	}
+	if a.Resource == nil {
+		return ErrNilResource
+	}
+	if a.Resource.ProviderType != pc.Type {
+		return ErrProviderTypeMismatch
+	}
+	if a.Resource.ProviderURN != pc.URN {
+		return ErrProviderURNMismatch
+	}
+	return nil
+}
+
+func getPermissions(resourceConfigs []*domain.ResourceConfig, a *domain.Appeal) ([]Permission, error) {
```

Comment:
let's update this to retrieve `permissions` from appeal instead of resource config

Reply:
done in https://github.com/odpf/guardian/pull/227/commits/3718c326b7f07d0e578375986ac9e6a39084197d

Reply:
can we close this conversation? @rahmatrhd